URL management topic update (wildcards) (#1070)
ezsystems/developer-documentation: docs/guide/url_management.md

# URL management

> Manage external URL addresses and URL wildcards with a comfortable user interface that is available in the Back Office, Admin tab, the URL Manager node. Configure URL aliases to have human-readable URL addresses throughout your system.

Author: I thought the features called for a bit of introduction. The only explanation on what URL management is can be found in the release notes, which are not permanent.
That said, I would rather leave the intro here.

> Manage external URL addresses and URL wildcards with a comfortable user interface that is available in the Back Office, Admin tab, the URL Manager node. Configure URL aliases to have human-readable URL addresses throughout your system.

Reviewer: Fair enough, but you will need to duplicate/repeat the same information in the user guide as well. I'd suggest a bit more adaptation:
```suggestion
You can manage external URL addresses and URL wildcards in the Back Office, **Admin** tab, the **URL Manager** node. Configure URL aliases to have human-readable URL addresses throughout your system.
```

## Link manager

When developing a site, users can enter links to external websites in either RichText or URL fields.  Each such link is then stored in the URL table. You can view and update all external links that exist within the site, without having to modify and re-publish the individual Content items.

The Link manager tab contains all the information about each link, including its status (valid or invalid) and the time the system last attempted to validate the URL address. Click an entry in the list to display its details and check which Content items use this link. Edit the entry to update the URL address in all the occurrences throughout the website.

!!! note

    When you edit the details of an entry to update the URL address, the status automatically changes to valid.

## External URL validation

> You can manually validate all the addresses from the URL table, by executing the `ezplatform:check-urls` command. It validates the links by accessing them one by one and updates the value in the Last checked field. If a broken link is found, its status is set to "invalid".

Reviewer: ```suggestion
You can validate all the addresses from the URL table by executing the `ezplatform:check-urls` command. It validates the links by accessing them one by one and updates the value in the Last checked field. If a broken link is found, its status is set to "invalid".
```
The same command can be run automatically, so I'd get rid ot the "manually"

Author: Automatic validation is mentioned below, hence the word "manually" to underline the difference between the two processes.

The following protocols are currently supported:

- `http`
- `https`
- `mailto`

### Enabling automatic URL validation

> To enable automatic URL validation, you set up cron to run the `ezplatform:check-urls` command periodically.

Reviewer: ```suggestion
To enable automatic URL validation, set up cron to run the `ezplatform:check-urls` command periodically.
```

# URL management

> You can manage external URL addresses and URL wildcards in the Back Office, **Admin** tab, the **URL Manager** node.

Reviewer: ```suggestion
You can manage external URL addresses and URL wildcards in the Back Office, **Admin** tab, the **URL Management** node.
```

                    	enabled: false
```

> Available options are protocol-specific. For details, see the tables below.

Reviewer: Not really necessary, as the tables are immediately below this.

#### http/https protocol

| Option             | Description                                                                              |  Default value |
|----------------------|----------------------------------------------------------------------------------------|-----------------------|
| enabled            | Enables   link validation.                                                      | true          |
| timeout            | Defines the time that the request is allowed to take (in seconds).                       | 10            |
| connection_timeout | Defines the time that the connect phase is allowed to take (in seconds).                 | 5             |
| batch_size         | Defines a maximum   number of asynchronous requests.                                     | 10            |

> | ignore_certificate | Decides if the peer's SSL certificate or the certificate's is verified against the host. | false         |

Reviewer: ```suggestion
| ignore_certificate | Decides if the peer's SSL certificate or the certificate name are verified against the host. | false         |
```

|----------------------|----------------------------------------------------------------------------------------|-----------------------|
| enabled            | Enables link validation.                                                      | true          |
| timeout            | Defines the time that the request is allowed to take (in seconds).                       | 10            |
| connection_timeout | Defines the time that the connect phase is allowed to take (in seconds).                 | 5             |
| batch_size         | Defines a maximum number of asynchronous requests.                                     | 10            |
| ignore_certificate | Decides if the peer's SSL certificate or the certificate name are verified against the host. | false         |

#### mailto protocol

| Option             | Description                                                         | Default value |
|--------------------|---------------------------------------------------------------------|---------------|
| enabled            | Enables link validation.                                            | true          |

> For more information about configuration, see [Configuration](configuration).

Reviewer 2: Note: This link is not valid now (404)

interface URLHandlerInterface
{
    /**

>      * Validates a given list of URLs.

Reviewer: ```suggestion
     * Validates given list of URLs.
```
This is part of product code

Author: Is it how the inline comments look in the product or is it an example with comments added for clarity?

Reviewer: This is an exact quote from code, see https://github.com/ezsystems/ezplatform-kernel/blob/master/eZ/Bundle/EzPublishCoreBundle/URLChecker/URLHandlerInterface.php#L12

        - { name: ezpublish.url_handler, scheme: custom }
```

The `scheme` attribute is mandatory and has to correspond to the name of the protocol, for instance, `ftp`.

## URL wildcards

With wildcards you have a simple way to display content under a specific URL address, for example, when you decide to reorganize the architecture and want to protect users from losing easy access to the content that they have bookmarked.

> For each wildcard you can decide whether the user will see the updated content at the original address (direct type), or will be redirected to the new address (forward type).

Reviewer: ```suggestion
For each wildcard you can decide whether the user sees the updated content at the original address (Direct type), or is redirected to the new address (Forward type).
```

You can make use of pre-defined transformation groups.
You can also add your own, with your own set of commands.
To add commands to an existing group, provide the group name and list the commands that you want to add.

### Regenerating URL aliases

You can use the `ezplatform:urls:regenerate-aliases` command to regenerate all URL aliases.
After the command is applied, old aliases redirect to the new ones.

Use it when:

- you change URL alias configuration and want to regenerate old aliases
- you encounter database corruption
- you have content that does not have a URL alias

!!! caution

    Before you apply the command, back up your database and make sure it is not modified while the command is running.

Execute the following command to regenerate aliases:

``` bash
bin/console ezplatform:urls:regenerate-aliases
```

You can also extend the command with the following parameters:

- `--iteration-count` — Defines how many Locations are processed at once to reduce memory usage
- `--location-id` — Regenerates URL addresses for specific Locations only, e.g. `ezplatform:urls:regenerate-aliases --location-id=1 --location-id=2`

## URL wildcards

> With wildcards, you have a simple way to change the URL address for many content items at the same time, by replacing a portion of the destinaton's URL address.

Reviewer: ```suggestion
With wildcards, you can change the URL address for many Content items at the same time, by replacing a portion of the destination's URL address.
```

For example, you might want to shorten the path, or make the path meaningful.

For each wildcard you can decide whether the user sees the content at the address that uses wildcards (Direct type), or is redirected to the original URL address of the destination (Forward type).

For example, a URL wildcard called `pictures/*/*` can use `media/images/{1}/{2}` as destination.
In this case, accessing `<yourdomain>/pictures/home/photo/` loads `<yourdomain>/media/images/home/photo/`.

You can configure URL wildcards either in the Back Office, or with the Public API.

Before you configure URL wildcards, you must enable the feature in configuration in the `config/packages/ezplatform.yaml` file:

``` yaml
ezplatform:
    url_wildcards:
        enabled: true
```

### Configuring URL wildcards in the Back Office

The URL wildcards tab contains all the information about each URL wildcard. You can delete or modify existing entries, or create new ones.

When you edit an existing wildcard URL or add a new one, you define the wildcard pattern and its destination.

> Also, you decide whether the user should see the updated content at the existing address (direct type), or should be redirected to the new address (forward type).

Reviewer: This information is repeated
